benchmarks/tests/md_mod.py

```diff
@@ -57,8 +57,6 @@ def compute ( p_num: int, d_num: int, pos: 'double[:,:]', vel: 'double[:,:]',
 
     kinetic = 0.5 * mass * kinetic
 
-    return potential, kinetic
```

Comment:
This is removed. But the arrays are still calculated. Shouldn't the calculation also be removed given that we don't use the result?

Reply:
@yguclu Proposed solution to remove code whose result is ignored : bca47a4

Reply:
I did not notice this removal. Every test in our benchmark suite should return some kind of output, in order to allow the user/developer to check the correctness of the accelerated code.

Reply:
@jalalium Didn't we decide that the test should return the total potential and kinetic energy at the end of the simulation?

Reply:
The md test does return the potential and kinetic energy at the end of the simulation. The `compute` function is called at each timestep.

Previously the energy was calculated, returned and ignored. It was changed so it was calculated but not returned, then recalculated in `test_md` after the simulation and the final energies are returned.

This commit removes the calculation inside the timestep whose result is unused

Reply:
That said, in order to return the energy from `test_md` it seems that the compute function has simply been duplicated. Either the force update should be removed from `test_md` or we should revert to the original organisation but also return the energy from `md`

Reply:
If you see a solution for avoiding useless calculations at each time step and code duplication, please go ahead!

Reply:
That is what I did in , I also added 0ac85cb to further reduce code duplication. But there is still some duplication between `compute` and `test_md`.

I have now added 7a9951b which reorganises the file somewhat, but I think this is the cleanest option

Reply:
👍 It looks good indeed!

Reply:
> @jalalium Didn't we decide that the test should return the total potential and kinetic energy at the end of the simulation?

It does in the main test function, returning it from the `calculate` function itself raised an error in `Fortran`

Reply:
What kind of error? Does the new file compile?

Reply:
Yes, your last update works fine in `Fortran` and `C` and gives the correct output

Reply:
but this version does not work with `Fortran`

```


# coding: utf-8
#------------------------------------------------------------------------------------------#
# This file is part of Pyccel which is released under MIT License. See the LICENSE file or #
# go to https://github.com/pyccel/pyccel/blob/master/LICENSE for full license details.     #
#------------------------------------------------------------------------------------------#
"""
Functions for running a small molecular dynamics simulation. The code is adapted from examples written by [J. Burkardt](https://people.sc.fsu.edu/~jburkardt/py_src/py_src.html)
To be accelerated with pyccel or pythran
"""
import sys
from numpy import zeros
from numpy import sqrt
from numpy import pi
from numpy import sin

# ================================================================
def compute ( p_num: int, d_num: int, pos: 'double[:,:]', vel: 'double[:,:]',
             mass: float, force: 'double[:,:]' ):
    """ Calculate the energy and forces associated with the current configuration
    """

    rij = zeros ( d_num )

    potential = 0.0

    for i in range ( 0, p_num ):
        #
        #  Compute the potential energy and forces.
        #
        for j in range ( 0, p_num ):
            if ( i != j ):
                #  Compute RIJ, the displacement vector.
                for k in range ( 0, d_num ):
                    rij[k] = pos[k,i] - pos[k,j]

                #  Compute D and D2, a distance and a truncated distance.
                d = 0.0
                for k in range ( 0, d_num ):
                    d = d + rij[k] ** 2

                d = sqrt ( d )
                d2 = min ( d, pi / 2.0 )

                #  Attribute half of the total potential energy to particle J.
                potential = potential + 0.5 * sin ( d2 ) * sin ( d2 )

                #  Add particle J's contribution to the force on particle I.
                for k in range ( 0, d_num ):
                    force[k,i] = force[k,i] - rij[k] * sin ( 2.0 * d2 ) / d
    #
    #  Compute the kinetic energy.
    #
    kinetic = 0.0
    for k in range ( 0, d_num ):
        for j in range ( 0, p_num ):
            kinetic = kinetic + vel[k,j] ** 2

    kinetic = 0.5 * mass * kinetic

    return potential, kinetic

# ================================================================
def update ( p_num: int, d_num: int, pos: 'double[:,:]', vel: 'double[:,:]',
            force: 'double[:,:]', acc: 'double[:,:]', mass: float, dt: float ):
    """ Update the position, velocity and force of the particles
    """

    rmass = 1.0 / mass
    #
    #  Update positions.
    #
    pos += vel * dt + 0.5 * acc * dt * dt
    #
    #  Update velocities.
    #
    vel += 0.5 * dt * ( force * rmass + acc )
    #
    #  Update accelerations.
    #
    acc[:] = force * rmass

# ================================================================
def r8mat_uniform_ab ( r: 'double[:,:]', m: int, n: int, a: float, b: float, seed: int ):
    """ Fill r with random numbers with a uniform distribution
    """

    i4_huge = 2147483647

    if ( seed < 0 ):
        seed = seed + i4_huge

    elif ( seed == 0 ):
        print ( '' )
        print ( 'R8MAT_UNIFORM_AB - Fatal error!' )
        print ( '  Input SEED = 0!' )
        sys.exit ( 'R8MAT_UNIFORM_AB - Fatal error!' )

    elif ( seed > 0 ):

        for j in range ( 0, n ):
            for i in range ( 0, m ):

                k = ( seed // 127773 )

                seed = 16807 * ( seed - k * 127773 ) - k * 2836

                seed = ( seed % i4_huge )

                if ( seed < 0 ):
                    seed = seed + i4_huge

                r[i,j] = a + ( b - a ) * seed * 4.656612875E-10

    return seed

# ================================================================
def initialize ( pos: 'double[:,:]', p_num: int, d_num: int ):
    """ Initialise the positions of the particles
    """
    #  Positions.
    seed = 123456789
    seed = r8mat_uniform_ab ( pos, d_num, p_num, 0.0, 10.0, seed )

# ================================================================
def md (d_num: int, p_num: int, step_num: int, dt: float,
        vel: 'double[:,:]', acc: 'double[:,:]',
        force: 'double[:,:]', pos: 'double[:,:]'):
    """ Run molecular dynamics simulation
    """

    mass = 1.0

    initialize ( pos, p_num, d_num )
    compute ( p_num, d_num, pos, vel, mass, force )
    
    for _ in range ( step_num - 1):
        update ( p_num, d_num, pos, vel, force, acc, mass, dt )
        compute ( p_num, d_num, pos, vel, mass, force )

    update ( p_num, d_num, pos, vel, force, acc, mass, dt )
    return compute ( p_num, d_num, pos, vel, mass, force )

# ================================================================
# pythran export test_md(int,int,int,float)
def test_md ( ):
    """ Run molecular dynamics test
    """

    #  Input
    d_num = 3
    p_num = 100
    step_num = 5000
    dt = 0.1
    #  Velocities.
    vel = zeros ( ( d_num, p_num ) )
    #  Accelerations.
    acc = zeros ( ( d_num, p_num ) )
    # Forces
    force = zeros ( ( d_num, p_num ) )
    # Positions
    pos = zeros ( ( d_num, p_num ) )

    return md(d_num, p_num, step_num, dt, vel, acc, force, pos)


if __name__ == '__main__':
    print(test_md())
```

It gives the following error:

```

  178 |       allocate(Out_0001_ptr(0:size(Out_0001_ptr, kind=i64) - 1_i64))
      |                           1
Error: ‘out_0001_ptr’ must not appear in the array specification at (1) in the same ALLOCATE statement where it is itself allocated
```

Reply:
I have opened https://github.com/pyccel/pyccel/issues/1321